### PR TITLE
Feature/a b graph [ PART 2 ]

### PR DESCRIPTION
**Pull request by pNagi**

### CODE REVIEW

**Comment by 0xchimz** [comment]
**ISSUE**
Each node in Apollo client state was differentiated by `_id`  but `_id` of nodeType was resolved by `actionTypeId` which is same for every version. That why apollo client state cannot differentiate each version of feature. 

**WHAT'S HAPPENING?**
1. Version `A` fetched a data from server including `nodes` which `_id` is an `actionTypeId`
2. Apollo set a new state by using received data.
3. Version `B` fetched a data from server including `nodes` which `_id` is an `actionTypeId`, same as version `A`.
4. Apollo overwrite version `A` 's node which has `_id` as same as `_id` of version `B` 's node. 

It's mean, feature `A` and feature `B` are using same element in a store.

**SUGGESTION**
Resolve `_id` of node by using a unique `_id` for each version. 

Example, `_id` = `actionTypeId + version.name`

**ResolveHelper.js** - pass `inputVersion` to each node
```javascript
export const generateNodes = (sessions = [], inputVersion) => _.flow([
  sessions => _.filter(sessions, filterByVersion(inputVersion)),
  sessions => _.map(sessions, mapSessionToNodes),
  nodes => _.flattenDeep(nodes),
  nodes => _.groupBy(nodes, node => node.actionTypeId),
  nodes => _.map(nodes, node => ({ ...node[0], count: node.length, inputVersion }))
])(sessions)
...
...
...
export const mapActionsToLinks = inputVersion => session => {
  const actions = removeFocusAndBlur(session.actions)
  const links = []

  if (actions.length > 1) {
    for (let i = 0; i < actions.length - 1; i++) {
      links.push({
        source: `${actions[i].actionTypeId}.${inputVersion.name}`,
        target: `${actions[i + 1].actionTypeId}.${inputVersion.name}`
      })
    }
  }
...
...
...
export const generateLinks = (sessions = [], inputVersion) => _.flow([
  sessions => _.filter(sessions, filterByVersion(inputVersion)),
  sessions => _.map(sessions, mapActionsToLinks(inputVersion)),
  links => _.flattenDeep(links),
  links => _.groupBy(links, link => link.source + link.target),
  links => _.map(links, link => ({ ...link[0], count: link.length, inputVersion })),
])(sessions)
...
...
...
export const generatePaths = (sessions = [], inputVersion) => _.flow([
  sessions => _.filter(sessions, filterByVersion(inputVersion)),
  sessions => _.map(sessions, mapSessionsToPaths),
  paths => _.groupBy(paths, reduceActionTypeIds),
  paths => _.map(paths, path => ({ ...path[0], count: path.length, inputVersion })),
])(sessions)
```

**PathType.js** - add `inputVersion` to each `action`
```javascript
const PathType = new GraphQLObjectType({
  name: 'Path',
  description: 'Path type',
  fields: () => ({
    _id: { type: new GraphQLNonNull(GraphQLString) },
    nodes: {
      type: new GraphQLList(NodeType),
      resolve: ({ actions, inputVersion }) => (
        _.map(actions, action => ({ ...action, inputVersion }))
      )
    },
    count: { type: new GraphQLNonNull(GraphQLInt) },
  })
```

**NodeType.js** - field `_id` should resolve with `actionTypeId` and `inputVersion.name`
```javascript
const NodeType = new GraphQLObjectType({
  name: 'Node',
  description: 'Node type',
  fields: () => ({
    _id: {
      type: new GraphQLNonNull(GraphQLString),
      resolve: root => (
        `${root.actionTypeId}.${root.inputVersion.name}`
      )
    },
    type: { type: new GraphQLNonNull(GraphQLString) },
    data: {
      type: new GraphQLNonNull(GraphQLString),
      resolve: root => root.url || removeInteractClick(root.target) || root.endpoint
    },
    count: { type: GraphQLInt },
  })
})
```

File: src/shared/nt-core/MainPage.styl [hunk 1]
```diff
@@ -24,9 +24,9 @@ $vertical-padding = 15px
   left 0
   z-index 1
   width $side-bar-size
-  background-color c-gray600
-  background linear-gradient(to top, c-gray850, c-gray800, c-gray800, c-gray850)
-  box-shadow 1px 0 5px 0 rgba(c-gray900, 0.5)
+  background-color c-gray850
+  // background linear-gradient(to top, c-gray850, c-gray800, c-gray800, c-gray850)
+  // box-shadow 1px 0 5px 0 rgba(c-gray900, 0.5)
```
Comment (0xchimz) [inline src/shared/nt-core/MainPage.styl]: remove a comment before merging with a master branch?

File: src/shared/nt-core/MainPage.styl [hunk 2]
```diff
@@ -37,7 +37,7 @@ $vertical-padding = 15px
   display flex
   justify-content center
   padding $vertical-padding $horizontal-padding
-  background-color c-gray25
+  // background-color c-gray25
```
Comment (0xchimz) [inline src/shared/nt-core/MainPage.styl]: remove a comment?

File: src/shared/nt-results/ResultsByLoadContainer.react.js [hunk 3]
```diff
@@ -84,37 +84,41 @@ class ResultsByLoadContainer extends React.Component {
...
+      this.setState({ x: this.container.clientWidth / 2 })
+    }
   }
 
   dragged = () => {
-    this.setState({ x: d3.event.x < 0 ? 0 : d3.event.x })
+    const x = d3.event.x
+    this.setState({
+      x: x < 0 ? 0 : x > this.container.clientWidth
+        ? this.container.clientWidth : x
```
Comment (0xchimz) [inline src/shared/nt-results/ResultsByLoadContainer.react.js]: Is this better?
```javascript
this.setState({
  x: (x < 0) ? 0 : (
    x > this.container.clientWidth ? (
      this.container.clientWidth
    ) : x
  )
})
```

File: src/shared/nt-results/ResultsPage.react.js [hunk 4]
```diff
@@ -19,6 +19,16 @@ class ResultsPage extends React.Component {
...
+      _id: PropTypes.string.isRequired,
+      name: PropTypes.string.isRequired,
+    })).isRequired,
+  }
+
+  state = { selectedFeatureId: undefined }
+
+  handleSelectFeature = featureId => () => {
+    this.setState({ selectedFeatureId: featureId })
   }
```
Comment (0xchimz) [inline src/shared/nt-results/ResultsPage.react.js:32]:
```javascript
handleSelectFeature = selectedFeatureId => () => {
    this.setState({ selectedFeatureId })
}
```

File: src/shared/nt-results/ResultsPage.react.js [hunk 5]
```diff
@@ -57,9 +66,20 @@ class ResultsPage extends React.Component {
...
           {this.renderLink('setting')}
+          Sessions
           {this.props.sessionTypes.map(sessionType => (
             this.renderLink(url.parse(sessionType.url).path, `/${sessionType._id}`)
           ))}
+          Features
+          {this.props.features.map(feature => (
+            <div
+              className={styles.nt__link}
+              key={feature._id}
```
Comment (0xchimz) [inline src/shared/nt-results/ResultsPage.react.js]: Should be `Link` or add cursor `pointer` to nt__link?

File: src/shared/nt-results/ResultsPage.styl [hunk 6]
```diff
@@ -24,10 +24,11 @@ $sidebar-size = 200px
     transition background-color 0.1s ease color 0.1s ease
```
Comment (0xchimz) [inline src/shared/nt-results/ResultsPage.styl:38]: Add `cursor: pointer;` ?

**Review by 0xchimz — CHANGES_REQUESTED** [review]

File: src/shared/nt-results/ResultsByLoadContainer.react.js (hunk 3 above)
Comment (pNagi) [reply src/shared/nt-results/ResultsByLoadContainer.react.js]: I don't think so, but I didn't mean what I write is good. Gonna look for better way to write this.

File: src/shared/nt-results/ResultsPage.react.js (hunk 4 above)
Comment (pNagi) [reply src/shared/nt-results/ResultsPage.react.js:32]: The `featureId` receive from params is not **selected** yet. It is **going to be selected**.

File: src/shared/nt-results/ResultsPage.react.js [hunk 7]
```diff
@@ -64,13 +65,11 @@ class ResultsPage extends React.Component {
...
     return (
       <div className={styles.nt}>
-        <div className={styles.nt__sidebar}>
-          {this.renderLink('setting')}
-          Sessions
+        <div className={styles.nt__nav}>
+          {this.renderLink('setting', 'cog')}
           {this.props.sessionTypes.map(sessionType => (
-            this.renderLink(url.parse(sessionType.url).path, `/${sessionType._id}`)
+            this.renderLink(url.parse(sessionType.url).path, '', `/${sessionType._id}`)
```
Comment (0xchimz) [inline src/shared/nt-results/ResultsPage.react.js]: if more than one argument of a method can be empty or null, you should pass arguments with an object type.

For example.
```javascript 
renderLink = (content, { iconName, path = '' }) => (
    <NavLink
      key={path}
      to={`${this.props.match.url}${path}`}
      activeClassName={styles.nt__activeLink}
      exact
    >
      <div className={styles.nt__link}>
        { iconName && <i className={`fa fa-${iconName}`} aria-hidden='true' />}
        <span>{content}</span>
      </div>
    </NavLink>
  )
....
...
...
this.renderLink('setting', { iconName: 'cog' })
....
this.renderLink(url.parse(sessionType.url).path, { path: `/${sessionType._id}` })
```

VSCode also have a guide when calling a method.
image

**Review by 0xchimz — CHANGES_REQUESTED** [review]

File: src/shared/nt-results/ResultsPage.react.js (hunk 7 above)
Comment (pNagi) [reply src/shared/nt-results/ResultsPage.react.js]: Right now this method is duplicate and need to be refactored in the future, so this problem will be gone with the refactoring. Anyways, I'm not sure if it still have time to do this.

Comment (0xchimz) [reply src/shared/nt-results/ResultsPage.react.js]: ok, both of them should be coded according to the suggestion above.

**Comment by 0xchimz** [comment]
image
I can't recognize this to be a button. I suggest adding a border or shadow.

**Review by 0xchimz — APPROVED** [review]

**Pull request merged**